Grbl_Esp32/src/Pin.cpp

```diff
@@ -0,0 +1,159 @@
+    auto nameStart = str.begin();
+    auto idx       = nameStart;
+    for (; idx != str.end() && *idx != '.' && *idx != ':'; ++idx) {
+        *idx = char(::tolower(*idx));
+    }
+    String prefix = str.substring(0, int(idx - str.begin()));
+
+    if (idx != str.end()) {  // skip '.'
+        ++idx;
+    }
+
+    int pinNumber = 0;
+    if (prefix != "") {
+        if (idx == str.end()) {
+            // Incorrect pin definition.
+            return false;
+        }
+
+        for (; idx != str.end() && *idx >= '0' && *idx <= '9'; ++idx) {
+            pinNumber = pinNumber * 10 + int(*idx - '0');
+        }
+        if (pinNumber < 0 || pinNumber > 253) {
+            // Pin number has to be between [0,253].
+            return false;
+        }
+    }
+
+    String options;
+    if (idx != str.end()) {
+        if (*idx != ':') {
+            // Pin definition attributes or EOF expected.
+            return false;
+        }
+        ++idx;
+
+        options = str.substring(int(idx - str.begin()));
+    }
+
+    // What would be a simple, practical way to parse the options? I figured, why not
+    // just use the C-style string, convert it to lower case, and change the separators
+    // into 'nul' tokens. We then pass the number of 'nul' tokens, and the first char*
+    // which is pretty easy to parse.
+
+    // Build an options parser:
+    Pins::PinOptionsParser parser(options.begin(), options.end());
+
+    // Build this pin:
+    if (prefix == "gpio") {
+        pinImplementation = new Pins::GPIOPinDetail(uint8_t(pinNumber), parser);
+    } else if (prefix == "i2s") {
+#ifdef ESP_32
+        pinImplementation = new Pins::I2SPinDetail(uint8_t(pinNumber), parser);
+#else
+        return false;  // not supported
+#endif
+    } else if (prefix == "void") {
+        // Note: having multiple void pins has its uses for debugging. Note that using
+        // when doing 'x == Pin::UNDEFINED' will evaluate to 'false' if the pin number
+        // is not 0.
+        pinImplementation = new Pins::VoidPinDetail(uint8_t(pinNumber));
+    }
+
+#if defined PIN_DEBUG && defined ESP32
+    pinImplementation = new Pins::DebugPinDetail(pinImplementation);
+#endif
+    return true;
+}
+
+Pin Pin::create(const String& str) {
+    Pins::PinDetail* pinImplementation = nullptr;
+    try {
+#ifdef PIN_DEBUG
+#    ifdef ESP32
+        grbl_sendf(CLIENT_ALL, "Setting up pin: [%s]\r\n", str.c_str());
+#    endif
+#endif
+        if (!parse(str, pinImplementation)) {
+#ifdef ESP32
+            grbl_sendf(CLIENT_ALL, "Setting up pin: '%s' failed.", str.c_str());
+#endif
+            return Pin::UNDEFINED;
+        } else {
+            // Check if we already have this pin:
+            auto existingPin = Pins::PinLookup::_instance.FindExisting(pinImplementation);
+
+            // If we already had it, and we didn't find itself, remove the new instance:
+            if (existingPin >= 0) {
+#ifdef ESP32
```

Comment:
maybe needs to be PIN_DEBUG

Reply:
Both. You might not have noticed, but the Pin classes have unit tests that are compatible with Google test and ESP32 test. Google test runs on the x64 machine, so any grbl.h code needs to go out when it compiles that.